### PR TITLE
Different allow apps for Wifi/mobile network

### PR DESCRIPTION
**Pull request by beimoe**
Added a feature enable switch and icons for better understanding.

### CODE REVIEW

**Comment by schwabe** [comment]
Please just update existing PRs by force pushing instead of closing them and opening new ones. The notes that I left in #1445 still apply. The current form still has the problem that if a user flaots from mobile to wifi or vice versa the setting are kept which is a weird/unexpected behaviour from a user perspective.

**Comment by beimoe** [comment]
I use it for weeks and the behaviour is as follows:
Connected via WiFi -> WiFi "Allow List"
Wifi is lost and WLAN connection is made -> OpenVPN reconnects (due to interface change?)
Connecting via WLAN --> WLAN "Allow List"

The other way round is the same behaviour.

Could you check if this is the same behaviour on your side?

**Comment by schwabe** [comment]
OpenVPN allows to float a connectiong between different interfaces without reconnection. Maybe you are using some server/VPN service that does not support this but for me with OpenVPN 2.x it just continues the connection on interface change without reconnect.

File: main/src/main/res/drawable/ic_baseline_wifi_24.xml [hunk 1]
```diff
@@ -0,0 +1,9 @@
+<vector xmlns:android="http://schemas.android.com/apk/res/android"
+    android:width="24dp"
+    android:height="24dp"
+    android:viewportWidth="24"
+    android:viewportHeight="24">
+  <path
+      android:fillColor="@android:color/black"
+      android:pathData="M1,9l2,2c4.97,-4.97 13.03,-4.97 18,0l2,-2C16.93,2.93 7.08,2.93 1,9zM9,17l3,3 3,-3c-1.65,-1.66 -4.34,-1.66 -6,0zM5,13l2,2c2.76,-2.76 7.24,-2.76 10,0l2,-2C15.14,9.14 8.87,9.14 5,13z"/>
+</vector>
```
Comment (schwabe) [inline main/src/main/res/drawable/ic_baseline_wifi_24.xml:9]: Where do these icons come from and what is their license?

File: main/src/main/java/de/blinkt/openvpn/core/OpenVPNService.java [hunk 2]
```diff
@@ -1010,6 +1013,24 @@ private void addLocalNetworksToRoutes() {
...
+                return false;
+            } else if (activeNetwork.getType() == ConnectivityManager.TYPE_MOBILE) {
+                // "Mobile data enabled";
+                return true;
+            }
+        } else {
+            // "No internet is available";
+            return false;
+        }
+        return false;
```
Comment (schwabe) [inline main/src/main/java/de/blinkt/openvpn/core/OpenVPNService.java:1031]: If everything that is not mobile is considered wifi anyway, this whole block can be simplified. It should also be made clear in other parts of the app that this is the behaviour.

File: main/src/main/java/de/blinkt/openvpn/core/OpenVPNService.java [hunk 3]
```diff
@@ -1033,7 +1054,16 @@ private void setAllowedVpnPackages(Builder builder) {
             }
         }
 
-        for (String pkg : mProfile.mAllowedAppsVpn) {
+        HashSet<String> mAllowedAppsVpn;
+
+        if ( getConnectivityByMobile(this) && mProfile.mAllowAppMobile) {
+            mAllowedAppsVpn = mProfile.mAllowedAppsVpnMobile;
+            VpnStatus.logInfo("Mobile data is active.");
```
Comment (schwabe) [inline main/src/main/java/de/blinkt/openvpn/core/OpenVPNService.java:1061]: That log message in itself is not very useful. I am sure it was useful for debugging.

File: main/src/main/res/values/strings.xml [hunk 4]
```diff
@@ -336,6 +336,7 @@
     <string name="vpn_disallow_radio">VPN is used for all apps but exclude selected</string>
     <string name="vpn_allow_radio">VPN is used for only for selected apps</string>
     <string name="vpn_allow_bypass">Allow apps to bypass the VPN</string>
+    <string name="enable_allow_mobile">Enable extra setting for mobile data</string>
```
Comment (schwabe) [inline main/src/main/res/values/strings.xml:339]: I don't think  "extra setting for mobile data" is a good wording. It is heard to understand what that does.

File: main/src/ui/java/de/blinkt/openvpn/fragments/PackageAdapter.kt [hunk 5]
```diff
@@ -142,6 +144,26 @@ internal class PackageAdapter(c: Context, vp: VpnProfile) : RecyclerView.Adapter
         vpnAllowBypassSwitch.setOnCheckedChangeListener { _, isChecked -> mProfile.mAllowAppVpnBypass = isChecked }
 
         vpnAllowBypassSwitch.isChecked = mProfile.mAllowAppVpnBypass
+
+
+
+
```
Comment (schwabe) [inline main/src/ui/java/de/blinkt/openvpn/fragments/PackageAdapter.kt:150]: Why so much extra white space?

File: main/src/ui/java/de/blinkt/openvpn/fragments/PackageAdapter.kt [hunk 6]
```diff
@@ -142,6 +144,26 @@ internal class PackageAdapter(c: Context, vp: VpnProfile) : RecyclerView.Adapter
...
+        val vpnAllowMobile = settingsView.findViewById<View>(R.id.enable_allow_mobile) as Switch
+
+        vpnAllowMobile.setOnCheckedChangeListener { _, isChecked ->
+            mProfile.mAllowAppMobile = isChecked
+            filter.filter("")
```
Comment (schwabe) [inline main/src/ui/java/de/blinkt/openvpn/fragments/PackageAdapter.kt:155]: Why does selecting a mobile setting reset the filter?

File: main/src/ui/java/de/blinkt/openvpn/fragments/PackageAdapter.kt [hunk 7]
```diff
@@ -157,18 +179,35 @@ internal class PackageAdapter(c: Context, vp: VpnProfile) : RecyclerView.Adapter
...
+        viewHolder.checkBoxMobile.tag = mInfo.packageName
+        viewHolder.checkBoxMobile.setOnCheckedChangeListener(this)
 
         viewHolder.checkBox.isChecked = mProfile.mAllowedAppsVpn.contains(mInfo.packageName)
+        viewHolder.checkBoxMobile.isChecked = mProfile.mAllowedAppsVpnMobile.contains(mInfo.packageName)
+
+        if ( mProfile.mAllowAppMobile )
+            viewHolder.checkBoxMobile.visibility = CompoundButton.VISIBLE
+
+        else
```
Comment (schwabe) [inline main/src/ui/java/de/blinkt/openvpn/fragments/PackageAdapter.kt:191]: Dubious extra white space

File: main/src/ui/java/de/blinkt/openvpn/fragments/PackageAdapter.kt [hunk 8]
```diff
@@ -157,18 +179,35 @@ internal class PackageAdapter(c: Context, vp: VpnProfile) : RecyclerView.Adapter
...
     }
 
 
     override fun onCheckedChanged(buttonView: CompoundButton, isChecked: Boolean) {
         val packageName = buttonView.tag as String
-        if (isChecked) {
-            mProfile.mAllowedAppsVpn.add(packageName)
-        } else {
-            mProfile.mAllowedAppsVpn.remove(packageName)
+        if ( buttonView.id == R.id.app_selected ) {
```
Comment (schwabe) [inline main/src/ui/java/de/blinkt/openvpn/fragments/PackageAdapter.kt:198]: The project uses the default android style that doesn't have extra space after ( in if

File: main/src/ui/res/layout/allowed_application_settings.xml [hunk 9]
```diff
@@ -48,18 +49,101 @@
...
+            android:layout_width="0dip"
+            android:layout_gravity="fill_horizontal|top"
+            android:textAppearance="?android:attr/textAppearanceSmall"
+            android:textAlignment="viewStart" /> -->
+
+        <!-- <TextView
+        android:id="@+id/app_disabled"
+        android:layout_marginStart="8dip"
+        android:layout_gravity="top"
+        android:textAppearance="?android:attr/textAppearanceSmall" /> -->
```
Comment (schwabe) [inline main/src/ui/res/layout/allowed_application_settings.xml:119]: Please don't add commented out code.

**Comment by schwabe** [comment]
The submitter did not respond or resolve the open question. Closing this PR for now.

**Pull request closed without merging**